File: MumbleLink/src/main/java/jk_5/nailed/mumblelink/MumbleLink.java
package jk_5.nailed.mumblelink;

import cpw.mods.fml.common.FMLCommonHandler;
import cpw.mods.fml.common.Mod;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;
import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.common.gameevent.TickEvent;
import net.minecraft.client.Minecraft;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.Vec3;

/**
 * No description given
 *
 * @author jk-5
 */
@Mod(modid = "NailedMumble", name = "NailedMumble", version = "0.1", dependencies = "required-after:Nailed")
public class MumbleLink {

    private static final long notificationDelay = 100;
    private boolean messagePrinted = false;
    private long start = -1;

    @Mod.EventHandler
    public void preInit(FMLPreInitializationEvent event){
        Mumble.tryInit();

        FMLCommonHandler.instance().bus().register(this);
    }

    @SubscribeEvent
    public void onRender(TickEvent.RenderTickEvent event){
        Minecraft mc = Minecraft.getMinecraft();
        if(event.phase == TickEvent.Phase.START){
            if(!Mumble.isInited() && !Mumble.tryInit()){
                return;
            }
            this.update(mc);
            if(Mumble.isInited()){
                if(mc != null && mc.theWorld != null){
                    if(!this.messagePrinted) {
                        long now = mc.theWorld.getWorldTime();
                        if(this.start == -1) {
                            this.start = now;
                        }
                        if(start + notificationDelay < now){
                            this.messagePrinted = true;
                            mc.thePlayer.addChatMessage(new ChatComponentText("Linked to mumble"));
                        }
                    }
                }else{
                    this.messagePrinted = false;
                }
            }
        }
    }

    public void update(Minecraft mc){
        try{
            float fAvatarFrontX = 1;
            float fAvatarFrontY = 0;
            float fAvatarFrontZ = 1;

            float fCameraFrontX = 1;
            float fCameraFrontY = 0;
            float fCameraFrontZ = 1;

            float fAvatarTopX = 0;
            float fAvatarTopY = 1;
            float fAvatarTopZ = 0;

            float fCameraTopX = 0;
            float fCameraTopY = 1;
            float fCameraTopZ = 0;

            Vec3 camera = mc.thePlayer.getLookVec();

            //Position of the player
            float[] fAvatarPosition = {
                    Float.parseFloat(Double.toString(mc.thePlayer.posX)), // TODO: losing precision here
                    Float.parseFloat(Double.toString(mc.thePlayer.posZ)), // TODO: losing precision here
                    Float.parseFloat(Double.toString(mc.thePlayer.posY))}; // TODO: losing precision here

            //Unit vector pointing out of the players eyes (here Front looks into scene).
            float[] fAvatarFront = {
                    Float.parseFloat(Double.toString(camera.xCoord * fAvatarFrontX)), // TODO: losing precision here
                    Float.parseFloat(Double.toString(camera.zCoord * fAvatarFrontZ)), // TODO: losing precision here
                    Float.parseFloat(Double.toString(camera.yCoord * fAvatarFrontY))}; // TODO: losing precision here

            //Unit vector pointing out of the top of the avatars head (here Top looks straight up).
            float[] fAvatarTop = {fAvatarTopX, fAvatarTopZ, fAvatarTopY};

            float[] fCameraPosition = {
                    Float.parseFloat(Double.toString(mc.thePlayer.posX)), //TODO: losing precision here
                    Float.parseFloat(Double.toString(mc.thePlayer.posZ)), //TODO: losing precision here
                    Float.parseFloat(Double.toString(mc.thePlayer.posY))}; //TODO: losing precision here

            float[] fCameraFront = {
                    Float.parseFloat(Double.toString(camera.xCoord * fCameraFrontX)), //TODO: losing precision here
                    Float.parseFloat(Double.toString(camera.zCoord * fCameraFrontZ)), //TODO: losing precision here
                    Float.parseFloat(Double.toString(camera.yCoord * fCameraFrontY))}; //TODO: losing precision here

            float[] fCameraTop = {fCameraTopX, fCameraTopZ, fCameraTopY};

            String identity = mc.thePlayer.getGameProfile().getName();
            String context = "NailedGlobal";
            String name = "Nailed";
            String description = "Link plugin for Minecraft with ModLoader";

            Mumble.update(fAvatarPosition, fAvatarFront, fAvatarTop, name, description, fCameraPosition, fCameraFront, fCameraTop, identity, context);
        }catch(Exception ignored){

        }
    }
}

File: MumbleLink/src/main/scala/jk_5/nailed/mumblelink/MumbleLink.scala
package jk_5.nailed.mumblelink

import cpw.mods.fml.common.{FMLCommonHandler, Mod}
import net.minecraft.client.Minecraft.{getMinecraft => mc}
import cpw.mods.fml.common.gameevent.TickEvent
import net.minecraft.util.ChatComponentText
import cpw.mods.fml.common.event.FMLPreInitializationEvent
import cpw.mods.fml.common.eventhandler.SubscribeEvent

/**
 * No description given
 *
 * @author jk-5
 */
@Mod(modid = "NailedMumble", name = "NailedMumble", version = "0.1", dependencies = "required-after:Nailed", modLanguage = "scala")
object MumbleLink {

  private final val notificationDelay = 100l
  private var messagePrinted = false
  private var start = -1l

  @Mod.EventHandler def preInit(event: FMLPreInitializationEvent){
    Mumble.tryInit()
    FMLCommonHandler.instance.bus.register(this)
  }

  @SubscribeEvent def onRender(event: TickEvent.RenderTickEvent): Unit = event.phase match {
    case TickEvent.Phase.START =>
      if(!Mumble.inited && !Mumble.tryInit) return
      this.update()
      if(Mumble.inited){
        if(mc != null && mc.theWorld != null){
          if(!this.messagePrinted){
            val now = mc.theWorld.getWorldTime
            if(this.start == -1){
              this.start = now
            }
            if(start + notificationDelay < now){
              this.messagePrinted = true
              mc.thePlayer.addChatMessage(new ChatComponentText("Linked to mumble"))
            }
          }
        }else this.messagePrinted = false
      }
    case _ =>
  }

  def update() = try{
    val fAvatarFrontX = 1f
    val fAvatarFrontY = 0f
    val fAvatarFrontZ = 1f
    val fCameraFrontX = 1f
    val fCameraFrontY = 0f
    val fCameraFrontZ = 1f
    val fAvatarTopX = 0f
    val fAvatarTopY = 1f
    val fAvatarTopZ = 0f
    val fCameraTopX = 0f
    val fCameraTopY = 1f
    val fCameraTopZ = 0f
    val camera = mc.thePlayer.getLookVec
    val fAvatarPosition = Array(mc.thePlayer.posX.toFloat, mc.thePlayer.posZ.toFloat, mc.thePlayer.posY.toFloat)
    val fAvatarFront = Array(camera.xCoord.toFloat * fAvatarFrontX, camera.zCoord.toFloat * fAvatarFrontZ, camera.yCoord.toFloat * fAvatarFrontY)
    val fAvatarTop = Array(fAvatarTopX, fAvatarTopZ, fAvatarTopY)
    val fCameraPosition = Array(mc.thePlayer.posX.toFloat, mc.thePlayer.posZ.toFloat, mc.thePlayer.posY.toFloat)
    val fCameraFront = Array(camera.xCoord.toFloat * fCameraFrontX, camera.zCoord.toFloat * fCameraFrontZ, camera.yCoord.toFloat * fCameraFrontY)
    val fCameraTop = Array(fCameraTopX, fCameraTopZ, fCameraTopY)
    val identity = mc.thePlayer.getGameProfile.getName
    val context = "NailedGlobal"
    val name = "Nailed"
    val description = "Nailed mumble plugin"
    Mumble.update(fAvatarPosition, fAvatarFront, fAvatarTop, name, description, fCameraPosition, fCameraFront, fCameraTop, identity, context)
  }catch{
    case ignored: Exception =>
  }
}
